genie-agent/src/main/java/com/netflix/genie/agent/utils/EnvUtils.java

```diff
@@ -172,6 +180,52 @@ private EnvUtils() {
         return parsedEnvMap;
     }
 
+    /**
+     * Performs shell environment variables expansion on the given string.
+     *
+     * @param inputString a string, possibly containing shell variables
+     * @param environmentVariablesMap a map of string environment variables
+     * @return a new string with variables expanded
+     * @throws VariableSubstitutionException if the value for a variable appearing in input is not found in the map
+     */
+    public static String expandShellVariables(
+        final String inputString,
+        final Map<String, String> environmentVariablesMap
```

Comment:
So the problem here I could see eventually is that the `environmentVariablesMap` is unordered. A lot of our variable usage is cascading (e.g. `GENIE_JOB_DIRECTORY` ends up being used a lot or some env could be set by someone else in a setup script and we need it to be defined before others can be evaluated. We may need to use a linked hash map or something instead of a generic map?